taichi/transforms/auto_diff.cpp

```diff
@@ -46,6 +218,46 @@ class ConvertLocalVar : public BasicStmtVisitor {
   }
 };
 
+class ReverseOuterLoops : public BasicStmtVisitor {
+  using BasicStmtVisitor::visit;
+
+ private:
+  ReverseOuterLoops(const std::vector<Block *> &IB) : IB(IB) {
+  }
+
+  bool is_IB(Block *block) const {
+    return std::find(IB.begin(), IB.end(), block) != IB.end();
+  }
+
+  void visit(StructForStmt *stmt) {
+    loop_depth += 1;
+    if (!is_IB(stmt->body.get()))
+      stmt->body->accept(this);
+    loop_depth -= 1;
+  }
+
+  void visit(RangeForStmt *stmt) {
+    if (loop_depth >= 1) {
+      stmt->reversed = !stmt->reversed;
```

Comment:
I just thought about an issue that may have existed before this PR: since we (of course) modify `RangeForStmt::reversed` after the constructor, can `StmtFieldManager` handle the case and guarantee `irpass::same_statements` to be correct?

Reply:
Good point. Probably `StmtFieldManager` won't be correct. Please open an issue to track this.